### ISSUE TITLE
Discussion about softmax operating range

### ISSUE THREAD

**Issue by hyunwoongko** [NVIDIA/Megatron-LM#191]
@jaredcasper As a result of experiment, the operating range of fused softmax is too tight. Fused softmax can be used outside the current range of use. Any thoughts on changing this?

**Comment by hyunwoongko** [comment]
see this: https://github.com/tunib-ai/oslo/blob/v.1.1.2/oslo/fused_kernels_utils.py#L461

**Comment by jaredcasper** [comment]
The current version in Megatron supports up to 4096, I believe you can add more powers of two by extending the switch statement here: https://github.com/NVIDIA/Megatron-LM/blob/e156d2fea7fc5c98e645f7742eb86b643956d840/megatron/fused_kernels/scaled_masked_softmax.h#L567

**Comment by hyunwoongko** [comment]
I mean scalability at shorter lengths, not longer lengths.

**Comment by jaredcasper** [comment]
Can you be more specific? Do you mean sk < 16?

**Comment by hyunwoongko** [comment]
- OSLO side
I conducted some experiements, and I found many more cases where they could be fused inside the max length.
```python
    def is_available(dtype, bsz, np, sq, sk, use_triang_mask):
        if dtype != torch.half or sk > 2048 or sk <= 0:
            return False

        bsz_per_block = get_custom_kernels().get_batch_per_block(sq, sk, bsz, np)

        if use_triang_mask:
            if (
                sq == sk
                and (sk <= 64 or sk % 4 == 0)
                and (bsz * np) % bsz_per_block == 0
            ):
                return True

        else:
            if sq > 1 and sq % bsz_per_block == 0:
                return True

        return False
```

- this is megatron side.
```python
    def is_kernel_available(self, mask, b, np, sq, sk):
        attn_batches = b * np

        if (
            self.scaled_masked_softmax_fusion  # user want to fuse
            and self.input_in_float16  # input must be fp16
            and 16 < sk <= 4096  # sk must be 16 ~ 2048
            and sq % 4 == 0  # sq must be divisor of 4
            and attn_batches % 4 == 0  # np * b must be divisor of 4
        ):
            if 0 <= sk <= 4096:
                batch_per_block = self.get_batch_per_block(sq, sk, b, np)

                if self.attn_mask_type == AttnMaskType.causal:
                    if attn_batches % batch_per_block == 0:
                        return True
                else:
                    if sq % batch_per_block == 0:
                        return True
        return False
```

**Comment by jaredcasper** [comment]
I don't want to try to understand the differences between those code blocks. :) Can you be more specific about what cases you are asking about?

**Comment by hyunwoongko** [comment]
- shorter than sk < 16.
- when attn_batches is not diviser by 4.
- sq length has nothing to do with 4 when we used scale mask softmax (not upper triang)
- ...

**Comment by hyunwoongko** [comment]
I wrote that code two months ago, so I can't remember all the details, but it worked well under the conditions I made.

**Comment by jaredcasper** [comment]
I don't think fusion is going to be very beneficial in that small regime, so we haven't really thought about what would work in those cases.

**Comment by hyunwoongko** [comment]
If I share the results of my experiments on this and code, are you willing to reflect code changes?
I respect your decision as it's on the Megatron side.

Megatron has a fixed length when training, so it won't do much, 
but I've found this to be helpful in inference (especially text generation) situations.
Because, when inference, the length changes each time.

**Comment by jaredcasper** [comment]
If it helps with inference it'd be great if you could submit a PR with whatever experiment results you have. We'll try to take a look and merge it if it makes sense, but can't promise it'll be anytime soon since we're all kind of hosed with time more time-sensitive projects at the moment and will need to spend some time understanding the implications and what not before we can merge it. We do really appreciate you taking time to give back to the project, just may not have time to look at it for a while!
